DIP22 - private symbol (in)visibility

MartinNowak opened the pull request:
The main goal of [DIP22](http://wiki.dlang.org/DIP22) and this implementation is to resolve conflicts between private and public symbols. Adding a private symbol to a module or a class should not interfere w/ any code in another module.
This is achieved by ignoring private/package/protected symbols during lookup, thus turning protection from an access into a visibility check.
- This PR solves the main hurdle to fix 314, b/c enabling protection for selective/renamed import declarations creates many public/private symbol conflicts (e.g. between the public `writeln` symbol in `std.stdio` and the private alias `import std.stdio : writeln` in another module).
- The visibility of overloaded symbols with mixed protection is determined by the most visible symbol, but an additional access check is performed after overload resolution.
- Because dmd did previously miss some access checks, it was possible to use private symbols.
  In order to not break code by making those symbols invisible an additional lookup with a deprecation warning is performed if the previous scope searches didn't found any symbol.
- In order to also fix [Issue 3254](https://issues.dlang.org/show_bug.cgi?id=3254) it was necessary to disable the eager access error during lookup for overloadable symbols. This will for now turn a possibly incorrect error into a deprecation (see test/compilable/protection.d).
- This PR fixes `package` and `package(std)` protection, but also deprecates invisible symbol for now.

MartinNowak committed:
fix Issue 143

MartinNowak committed:
fix issue 1238

MartinNowak committed:
fix issue 1754

MartinNowak committed:
fix issue 2991

MartinNowak committed:
add test case for invisible private symbols in base class

MartinNowak committed:
test that public imports filter invisible symbols

- what symbols are visible depends on each module in an import chain so
  that anyone importing a module w/ a public import cannot see more than
  the module itself

MartinNowak committed:
visibility checks for dot identifier lookups

MartinNowak committed:
fix Issue 3254 - Module member visibility depends on declaration order

rsw0x commented:
Slightly off-topic, but does this DIP address internal linkage at all?

IIRC the spec makes statements about private symbols having the same linkage as static in C, but this is not true.

MartinNowak commented:
Linkage is completely off-topic, and no dmd always emits fully visible symbols, which isn't a problem due to D's mangling. We want to switch to visibility=hidden for non-exported symbols which allows further LTOs.

MartinNowak committed:
implement DIP22 - private symbol invisibility

- replace access checks by symbol invisibility
- no longer find private/package protected symbols
- resolves private vs. public symbol collisions in imported modules
- resolve mixed private/public overloads

MartinNowak committed:
rename IgnorePrivateMembers -> IgnorePrivateImports

- to clarify the function of this flag

WalterBright commented:
This seems to share a lot of the commits for https://github.com/D-Programming-Language/dmd/pull/5485 which makes it even more worthwhile to have each commit be a separate PR.

MartinNowak commented:
> This seems to share a lot of the commits for #5485 which makes it even more worthwhile to have each commit be a separate PR.

https://github.com/D-Programming-Language/dmd/pull/5485#issue-137096167 is based on this PR, so github also shows the commits of __this__ PR in the other PR.

MartinNowak commented:
> For example, the first commit is a simple refactor, there's no need to lump it in with the rest.

It's already hard enough to get anything reviewed. Why should I split off a trivial constant renaming to a separate PR, paying all the transaction costs again, and creating additional dependency issues for downstream maintainers or releases.
It's not like this is a big unrelated refactoring, you can click on the commit https://github.com/MartinNowak/dmd/commit/fd44e009ac658c1b4904359af2121c7a5ae96292, review it on it's own, and even put an OK comment on it if you can't review the complete PR in one go.

MartinNowak commented:
Github isn't the best tool for reviews (e.g. it show commits in the wrong order).
Those are the 2 core commits of this PR.
https://github.com/MartinNowak/dmd/commit/6c492ab8259e04dce8f663a3b3d11e5b01e43c5b
https://github.com/MartinNowak/dmd/commit/fbeb8fe25bdcf447224b06cc5a2e0c3e43620cc2

WalterBright commented:
> Why should I split off a trivial constant renaming to a separate PR, paying all the transaction costs again, and creating additional dependency issues for downstream maintainers or releases.

As I explained to Kenji:

1. trivial PRs are going to be reviewed a lot quicker, because they are much, much, much easier to review
2. encapsulated PRs make just as much sense as encapsulated code does. Having tangled code is as hard as tangled PRs
3. We have lots and lots of PRs that result in regressions. If the regression is caused by a smaller PR, it is a LOT easier to figure out what went wrong than a large PR
4. Having lots of stuff thrown into one PR means it's all or nothing. With a sequence of smaller PRs, one can reject one without rejecting all the rest of otherwise good stuff.
5. I know that some PRs will depend on previous PRs. I don't see that this is a big deal, just throw them up, and rebase as necessary.
6. Bug fixes should be focused, one per PR, and not have refactorings and other kitchen appliances thrown in.
7. Once a PR spans more than one screen, it gets a lot harder to review. Have you ever noticed that the more lines a PR has, the more the reviews consist of complaints about formatting and other trivia, instead of meat? I sure have, and I'm guilty of it as well.

> Those are the 2 core commits of this PR.

Please put the non-core ones as separate PRs.

MartinNowak commented:
Let's please stop wasting time and get to the matter. This PR isn't trivial and it can't be, but it's already simpler than your lookup change and already doesn't contain any unrelated changes.
The renaming of IgnorePrivateMember to IgnorePrivateImports (https://github.com/MartinNowak/dmd/commit/fd44e009ac658c1b4904359af2121c7a5ae96292) is part of this, b/c this PR changes search to ignore private symbols which collides with the name of the flag.
The other commits just add test cases of relevant bugs.

MartinNowak commented:
> Once a PR spans more than one screen, it gets a lot harder to review. Have you ever noticed that the more lines a PR has, the more the reviews consist of complaints about formatting and other trivia, instead of meat? I sure have, and I'm guilty of it as well.

Well, that's partly b/c github notifications are mostly gone once you click on them, so confronted with complex PRs and a lack of time, people use small comments as a TODO reminder. And if someone can't grasp a PR right away she should get into contact with the author, e.g. by leaving questions or by scheduling a Skype or IRC meeting if lots of discussion is necessary.
All of these are workflow issues and in any case it's a bad idea to disguise a complex change like this one behind many artificial and seemingly trivial changes.

MartinNowak commented:
This PR changes +118/-23 lines of non-test code. All it does is adding a new `symbolIsVisible(Module, Symbol)` function, uses that to ignore symbols during lookup, and performs another deprecated lookup when nothing was found.

andralex commented:
The way I see it is, the project lead is asking for some reasonable changes. Sure, packaging PRs is not an exact science, everybody has their quirks etc, so each and every thing asked can be debated. But at the top level: the project lead is asking for some reasonable changes. @MartinNowak, do you think there are ways to implement them so he can review and pull this PR?

MartinNowak commented:
> The way I see it is, the project lead is asking for some reasonable changes.

If you're referring to https://github.com/D-Programming-Language/dmd/pull/5472#issuecomment-190602217, this was a misunderstanding, #5485 is completely based on the PR, so of course both share the same commits. There isn't anything in this PR that isn't part of the DIP22 change.
I said in https://github.com/D-Programming-Language/dmd/pull/5472#issuecomment-191130614 why I had to rename a search flag before adding another flag and changing the search semantics.

WalterBright commented on src/dscope.d:
```diff
@@ -544,12 +544,27 @@ struct Scope
+         * checked by the compiler remain usable.  Once the deprecation is over,
+         * this should be moved to search_correct instead.
+         */
+        s = searchScopes(flags | SearchLocalsOnly | IgnoreSymbolVisibility);
```
Put check for IgnorErrors and useDeprecated==1 here to avoid unnecessary extra symbol lookups.

WalterBright replied:
This code may also need to be in expression.searchUFCS().

MartinNowak replied:
> Put check for IgnorErrors and useDeprecated==1 here to avoid unnecessary extra symbol lookups.

I'd still have to return the correct symbol, even if not reporting the deprecation warning.
I can try to gather some stats how often symbol lookup falls through to this old behavior.

MartinNowak replied:
> This code may also need to be in expression.searchUFCS().

Seems so, will add that and a test case this evening.

MartinNowak commented:
I'll run some benchmarks, but I don't think this affects anything b/c failed lookups are usually an error and in any case rare.
Comparing 2 test-runs already suggests that the timing didn't change.
https://auto-tester.puremagic.com/show-run.ghtml?projectid=1&runid=1976810&isPull=true
https://auto-tester.puremagic.com/show-run.ghtml?projectid=1&runid=1976778&isPull=true

MartinNowak committed:
also do a deprecated search for private symbols in searchUFCS

- for better error messages and to still allow any usage
  that might slip access check after overload resolution
- the better error message should be produced by search_correct
  once the deprecation is over

MartinNowak committed:
handle -transition=import/-transition=checkimports

- also do a deprecated search for private symbols in the new part of
  checkimports
- keep old behavior (ignoring symbol visibility) w/ -transition=import
  and the old part of -transition=checkimports

MartinNowak committed:
changelog entry for DIP22

MartinNowak commented:
I added 3 commits as you can see above.
- Add deprecated search for invisible symbols to `searchUFCS`. This is mostly improve error messages b/c most UFCS symbols should have access checks.
Still it's a lot more helpful to get `Deprecation: imports.dip22a.bar is not visible from module dip22a` instead of `Error: bar is not a property of ...`. Once the visibility deprecation is over we should add the fallback search for private symbols to the spell checker (`search_correct`).
- Integrate DIP22 with `-transition=import` and `-transition=checkimports`
- Describe all the changes in the changelog

MartinNowak commented:
Slowdown is very small 0.1-0.2% for phobos measured without object creation (`-o-`).

WalterBright commented:
> Slowdown is very small 0.1-0.2% for phobos measured without object creation

> without object creation

Not sure what that means.

Slow compile times are an accumulation of those "very small" .1-.2 slowdowns, like barnacles on a ship.

WalterBright commented:
Due to these slowdowns, I propose a shorter deprecation cycle for this.

MartinNowak commented:
> Due to these slowdowns, I propose a shorter deprecation cycle for this.

A few releases (2-3 / 4-6 month) should be fine, it's quite a mess to maintain these extra flags and lookups.

> Not sure what that means.

It mean I skipped the codegen pass by using `-o-`.

> Slow compile times are an accumulation of those "very small" .1-.2 slowdowns, like barnacles on a ship.

Well, 0.1% change in semantic speed shouldn't be an argument against such a big change.
We have quite a few bigger performance issues lingering around, e.g. the 10-15% slist_reset for static libraries (partly fixed now), ~4% in addDeferredSemantic3, or the pervasive use of lists instead of arrays.

WalterBright commented:
I know we have other time suckers, like the excessive template instantiations. We need to work on all of them, and try not to make things worse.

WalterBright commented:
Auto-merge toggled on

MartinNowak commented:
> We need to work on all of them

Yes, we need to work on all of them, but they are all fixable problems.
Just figured that the addDeferred is very low hanging fruit, so I fixed it #5506.

CyberShadow commented:
This pull request introduced a regression:
https://issues.dlang.org/show_bug.cgi?id=15785
